receiver/splunkhecreceiver/splunk_to_logdata.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package splunkhecreceiver
+
+import (
+	"errors"
+
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/translator/conventions"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/common/splunk"
+)
+
+const (
+	cannotConvertValue = "cannot convert field value to attribute"
+)
+
+// SplunkHecToLogData transforms splunk events into logs
+func SplunkHecToLogData(logger *zap.Logger, events []*splunk.Event) (pdata.LogSlice, error) {
+	lrs := pdata.NewLogSlice()
+	lrs.Resize(len(events))
+
+	for i, event := range events {
+		lr := lrs.At(i)
+		lr.InitEmpty()
+
+		// The SourceType field is the most logical "name" of the event.
+		lr.SetName(event.SourceType)
+		lr.Body().InitEmpty()
+		if value, ok := event.Event.(string); ok {
+			lr.Body().SetStringVal(value)
+		} else if value, ok := event.Event.(int64); ok {
+			lr.Body().SetIntVal(value)
+		} else if value, ok := event.Event.(float64); ok {
+			lr.Body().SetDoubleVal(value)
+		} else if value, ok := event.Event.(bool); ok {
+			lr.Body().SetBoolVal(value)
+		} else if value, ok := event.Event.(map[string]interface{}); ok {
+			mapValue, err := convertToAttributeMap(logger, value)
+			if err != nil {
+				return lrs, err
+			}
+			lr.Body().SetMapVal(mapValue)
+		} else if value, ok := event.Event.([]interface{}); ok {
+			arrValue, err := convertToArrayVal(logger, value)
+			if err != nil {
+				return lrs, err
+			}
+			lr.Body().SetArrayVal(arrValue)
+		}
+
+		// Splunk timestamps are in seconds so convert to nanos by multiplying
+		// by 1 billion.
+		lr.SetTimestamp(pdata.TimestampUnixNano(event.Time * 1e9))
+
+		attrs := lr.Attributes()
+		attrs.InitEmptyWithCapacity(3 + len(event.Fields))
+		attrs.InsertString(conventions.AttributeHostHostname, event.Host)
```

Comment:
Host and service name need to go in the Resource of this ResourceLogs instance, so all logs under the resourceLogs have to have the same resource, which implies some amount of grouping required.

Reply:
OK, this is done - I think.